java-manta-client-unshaded/src/main/java/com/joyent/manta/client/LazyMantaClient.java

```diff
@@ -0,0 +1,299 @@
+    @Override
+    HttpHelper getHttpHelper() {
+        ensureClientNotClosed();
+
+        if (reload.contains(MantaClientComponent.AUTH)
+                || reload.contains(MantaClientComponent.HTTP)) {
+            clearHttpHelper();
+            auth.reload();
+            reload.remove(MantaClientComponent.AUTH);
+            reload.remove(MantaClientComponent.HTTP);
+        }
+
+        if (lazyHttpHelper == null) {
+            lazyHttpHelper = buildHttpHelper();
+        }
+
+        return lazyHttpHelper;
+    }
+
+    @Override
+    UriSigner getUriSigner() {
+        ensureClientNotClosed();
+
+        if (reload.contains(MantaClientComponent.SIGN)) {
+            lazyUriSigner = null;
+            auth.reload();
+            reload.remove(MantaClientComponent.SIGN);
+        }
+
+        if (lazyUriSigner == null) {
+            lazyUriSigner = new UriSigner(getContext(), auth.getKeyPair(), auth.getSigner());
+        }
+
+        return lazyUriSigner;
+    }
+
+    // INTERNAL STATE MANAGEMENT
+
+    /**
+     * Clean up the existing {@link HttpHelper} so a new one can be created. Sets the helper reference
+     * to {@code null} so that the {@link AuthenticationConfigurator}'s
+     * {@link com.joyent.http.signature.ThreadLocalSigner} can be garbage-collected.
+     */
+    private void clearHttpHelper() {
+        if (lazyHttpHelper == null) {
+            return;
+        }
+
+        try {
+            lazyHttpHelper.close();
+            lazyHttpHelper = null;
+        } catch (final Exception e) {
+            throw new MantaLazyConfigurationException("Error occurred while closing existing HttpHelper", e);
+        }
+    }
+
+    /**
+     * Prepare a new {@link HttpHelper}, potentially passing a provided {@link MantaConnectionFactoryConfigurator}.
+     * Also registers the {@link MantaConnectionFactory} with {@link #lazyBeanSupervisor}.
+     *
+     * @return the {@link HttpHelper} to used based on new configuration
+     */
+    @SuppressWarnings("AvoidInlineConditionals")
+    private HttpHelper buildHttpHelper() {
+        final MantaConnectionFactory connectionFactory = connectionConfig != null
+                ? new MantaConnectionFactory(getContext(), auth.getKeyPair(), auth.getSigner(), connectionConfig)
+                : new MantaConnectionFactory(getContext(), auth.getKeyPair(), auth.getSigner());
+        final MantaApacheHttpClientContext connectionContext = new MantaApacheHttpClientContext(connectionFactory);
+
+        lazyBeanSupervisor.expose(connectionFactory);
+
+        if (BooleanUtils.isTrue(getContext().isClientEncryptionEnabled())) {
+            return new EncryptionHttpHelper(connectionContext, getContext());
+        }
+
+        return new StandardHttpHelper(connectionContext, getContext());
+    }
+
+    // LIFECYCLE METHODS
+
+    /**
+     * Closed clients have cleaned up resources and should no longer be used.
+     */
+    private void ensureClientNotClosed() {
+        if (isClosed()) {
```

Comment:
Please give the user of the SDK more information so they know what they did wrong and/or what were the circumstances that led them to this exception.